File: app-presentation/src/main/java/com/zxjdev/smile/presentation/infrastucture/splash/SplashActivity.java
package com.zxjdev.smile.presentation.infrastucture.splash;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.content.Intent;
import android.graphics.Typeface;
import android.os.Bundle;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationSet;
import android.view.animation.AnimationUtils;
import android.view.animation.RotateAnimation;
import android.view.animation.TranslateAnimation;
import android.widget.Button;

import com.zxjdev.smile.R;
import com.zxjdev.smile.presentation.authorization.login.LoginActivity;
import com.zxjdev.smile.presentation.authorization.register.RegisterActivity;
import com.zxjdev.smile.presentation.common.base.activity.ActivityModule;
import com.zxjdev.smile.presentation.common.base.activity.DaggerActivity;
import com.zxjdev.smile.presentation.infrastucture.main.MainActivity;
import com.zxjdev.smile.presentation.infrastucture.splash.di.SplashActivityComponent;
import com.zxjdev.smile.presentation.infrastucture.splash.di.SplashActivityModule;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * 开屏界面
 */
public class SplashActivity extends DaggerActivity implements SplashContract.View {

  @BindView(R.id.layout_button_container) View buttonContainer;
  @BindView(R.id.btn_login) Button btnLogin;
  @BindView(R.id.btn_register) Button btnRegister;

  @Inject SplashContract.Presenter presenter;

  private SplashActivityComponent splashActivityComponent;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    setContentView(R.layout.activity_splash);
    ButterKnife.bind(this);

    initUi();

    presenter.handleAutoLogin();
  }

  @Override
  protected void onStart() {
    super.onStart();
    // doButtonsPropertyAnimationFromCode();
  }

  @Override
  protected void onResume() {
    super.onResume();
    // doButtonsPropertyAnimationFromCode();
  }

  @Override
  protected void initDaggerComponent() {
    splashActivityComponent = getApplicationComponent().getSplashActivityComponent(new ActivityModule(this),
      new SplashActivityModule(this));
    splashActivityComponent.inject(this);
  }

  @Override
  protected void releaseDaggerComponent() {

  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    presenter.destroy();
    splashActivityComponent = null;
  }

  private void initUi() {
    // 设置自定义的字体
    Typeface customFont = Typeface.createFromAsset(this.getAssets(), "fonts/AaPangYaer.ttf");
    btnLogin.setTypeface(customFont);
    btnRegister.setTypeface(customFont);
  }

  @OnClick(R.id.btn_login)
  public void loginCLick() {
    navigateToLogin();
  }

  @OnClick(R.id.btn_register)
  public void registerClick() {
    navigateToRegister();
  }

  private void navigateToLogin() {
    Intent login = new Intent(context, LoginActivity.class);
    startActivity(login);
  }

  private void navigateToRegister() {
    Intent login = new Intent(context, RegisterActivity.class);
    startActivity(login);
  }

  @Override
  public void navigateToMain() {
    Intent intent = new Intent(context, MainActivity.class);
    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
    startActivity(intent);
  }

  @Override
  public void showButtons() {
    buttonContainer.setVisibility(View.VISIBLE);

    doButtonsViewAnimationFromRes();
    // buttonContainer.post(this::doButtonsPropertyAnimationFromCode);
    // buttonContainer.post(this::doButtonsViewAnimationFromCode);
  }

  private void doButtonsPropertyAnimationFromCode() {
    float btnLoginStartPosition = buttonContainer.getX() + buttonContainer.getMeasuredWidth();
    float btnLoginEndPosition = btnLogin.getX();

    ObjectAnimator btnLoginPosition = ObjectAnimator.ofFloat(btnLogin, "x", btnLoginStartPosition, btnLoginEndPosition);
    btnLoginPosition.setDuration(3000);

    ObjectAnimator btnLoginRotation = ObjectAnimator.ofFloat(btnLogin, "rotation", 0, -360);
    btnLoginRotation.setDuration(3000);

    float btnRegisterStartPosition = buttonContainer.getX() - btnRegister.getMeasuredWidth();
    float btnRegisterEndPosition = btnRegister.getX();
    ObjectAnimator btnRegisterPosition = ObjectAnimator.ofFloat(btnRegister, "x", btnRegisterStartPosition,
      btnRegisterEndPosition);
    btnRegisterPosition.setDuration(3000);

    ObjectAnimator btnRegisterRotation = ObjectAnimator.ofFloat(btnRegister, "rotation", 0, 360);
    btnRegisterRotation.setDuration(3000);

    AnimatorSet animatorSet = new AnimatorSet();
    animatorSet.playTogether(btnRegisterPosition, btnRegisterRotation, btnLoginPosition, btnLoginRotation);
    animatorSet.addListener(new Animator.AnimatorListener() {
      @Override
      public void onAnimationStart(Animator animation) {
      }

      @Override
      public void onAnimationEnd(Animator animation) {
      }

      @Override
      public void onAnimationCancel(Animator animation) {
      }

      @Override
      public void onAnimationRepeat(Animator animation) {

      }
    });
    animatorSet.start();
  }

  private void doButtonsViewAnimationFromCode() {
    float btnLoginStartPosition = -(btnLogin.getX() + btnLogin.getWidth());
    float btnLoginEndPosition = 0;

    TranslateAnimation btnLoginTranslate = new TranslateAnimation(Animation.ABSOLUTE, btnLoginStartPosition,
      Animation.ABSOLUTE, btnLoginEndPosition, Animation.RELATIVE_TO_SELF, 0, Animation.RELATIVE_TO_SELF, 0);

    RotateAnimation btnLoginRotate = new RotateAnimation(0, 360, Animation.RELATIVE_TO_SELF, 0.5f,
      Animation.RELATIVE_TO_SELF, 0.5f);

    AnimationSet btnLoginAnimation = new AnimationSet(true);
    btnLoginAnimation.addAnimation(btnLoginRotate);
    btnLoginAnimation.addAnimation(btnLoginTranslate);
    btnLoginAnimation.setDuration(3000);

    btnLogin.startAnimation(btnLoginAnimation);

    float btnRegisterStartPosition = buttonContainer.getWidth() - btnRegister.getX();
    float btnRegisterEndPosition = 0;

    TranslateAnimation btnRegisterTranslate = new TranslateAnimation(Animation.ABSOLUTE, btnRegisterStartPosition,
      Animation.ABSOLUTE, btnRegisterEndPosition, Animation.RELATIVE_TO_SELF, 0, Animation.RELATIVE_TO_SELF, 0);

    RotateAnimation btnRegisterRotate = new RotateAnimation(0, -360, Animation.RELATIVE_TO_SELF, 0.5f,
      Animation.RELATIVE_TO_SELF, 0.5f);

    AnimationSet btnRegisterAnimation = new AnimationSet(true);
    btnRegisterAnimation.addAnimation(btnRegisterRotate);
    btnRegisterAnimation.addAnimation(btnRegisterTranslate);
    btnRegisterAnimation.setDuration(3000);

    btnRegister.startAnimation(btnRegisterAnimation);
  }

  private void doButtonsViewAnimationFromRes() {
    Animation animRollInFromLeft = AnimationUtils.loadAnimation(getApplicationContext(), R.anim.anim_roll_in_from_left);
    Animation animRollInFromRight = AnimationUtils.loadAnimation(getApplicationContext(),
      R.anim.anim_roll_in_from_right);
    btnRegister.startAnimation(animRollInFromRight);
    btnLogin.startAnimation(animRollInFromLeft);
  }
}

File: app-presentation/src/main/java/com/zxjdev/smile/presentation/infrastucture/splash/SplashActivity.kt
@file:Suppress("unused")

package com.zxjdev.smile.presentation.infrastucture.splash

import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.content.Intent
import android.graphics.Typeface
import android.os.Bundle
import android.view.View
import android.view.animation.*
import com.zxjdev.smile.R
import com.zxjdev.smile.presentation.authorization.login.LoginActivity
import com.zxjdev.smile.presentation.authorization.register.RegisterActivity
import com.zxjdev.smile.presentation.common.base.activity.BaseActivity
import com.zxjdev.smile.presentation.infrastucture.main.MainActivity
import kotlinx.android.synthetic.main.activity_splash.*
import javax.inject.Inject

/**
 * 开屏界面
 */
class SplashActivity : BaseActivity(), SplashContract.View {

    @Inject internal lateinit var presenter: SplashPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_splash)
        initUi()

        presenter.takeView(this)
        presenter.handleAutoLogin()
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.dropView()
    }

    private fun initUi() {
        // 设置自定义的字体
        val customFont = Typeface.createFromAsset(this.assets, "fonts/AaPangYaer.ttf")
        btn_login.typeface = customFont
        btn_register.typeface = customFont
        btn_login.setOnClickListener({ navigateToLogin() })
        btn_register.setOnClickListener({ navigateToRegister() })
    }

    private fun navigateToLogin() {
        val login = Intent(context, LoginActivity::class.java)
        startActivity(login)
    }

    private fun navigateToRegister() {
        val login = Intent(context, RegisterActivity::class.java)
        startActivity(login)
    }

    override fun navigateToMain() {
        val intent = Intent(context, MainActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        startActivity(intent)
    }

    override fun showButtons() {
        layout_button_container.visibility = View.VISIBLE

        // doButtonsViewAnimationFromRes();
        layout_button_container.post { this.doButtonsPropertyAnimationFromCode() }
        // buttonContainer.post(this::doButtonsViewAnimationFromCode);
    }

    /**
     * 通过代码实现属性动画
     */
    private fun doButtonsPropertyAnimationFromCode() {
        val btnLoginStartPosition = layout_button_container.x + layout_button_container.measuredWidth
        val btnLoginEndPosition = btn_login.x

        val btnLoginPosition = ObjectAnimator.ofFloat(btn_login, "x", btnLoginStartPosition, btnLoginEndPosition)
        btnLoginPosition.duration = 3000

        val btnLoginRotation = ObjectAnimator.ofFloat(btn_login, "rotation", 0.toFloat(), (-360).toFloat())
        btnLoginRotation.duration = 3000

        val btnRegisterStartPosition = layout_button_container.x - btn_register.measuredWidth
        val btnRegisterEndPosition = btn_register.x
        val btnRegisterPosition = ObjectAnimator.ofFloat(btn_register,
                "x",
                btnRegisterStartPosition,
                btnRegisterEndPosition)
        btnRegisterPosition.duration = 3000

        val btnRegisterRotation = ObjectAnimator.ofFloat(btn_register, "rotation", 0.toFloat(), 360.toFloat())
        btnRegisterRotation.duration = 3000

        val animatorSet = AnimatorSet()
        animatorSet.playTogether(btnRegisterPosition, btnRegisterRotation, btnLoginPosition, btnLoginRotation)
        animatorSet.start()
    }

    /**
     * 通过代码实现View动画，按钮起始位置和结束位置是精确的
     */
    private fun doButtonsViewAnimationFromCode() {
        val btnLoginStartPosition = -(btn_login.x + btn_login.width)
        val btnLoginEndPosition = 0f

        val btnLoginTranslate = TranslateAnimation(Animation.ABSOLUTE,
                btnLoginStartPosition,
                Animation.ABSOLUTE,
                btnLoginEndPosition,
                Animation.RELATIVE_TO_SELF,
                0f,
                Animation.RELATIVE_TO_SELF,
                0f)

        val btnLoginRotate = RotateAnimation(0f,
                360f,
                Animation.RELATIVE_TO_SELF,
                0.5f,
                Animation.RELATIVE_TO_SELF,
                0.5f)

        val btnLoginAnimation = AnimationSet(true)
        btnLoginAnimation.addAnimation(btnLoginRotate)
        btnLoginAnimation.addAnimation(btnLoginTranslate)
        btnLoginAnimation.duration = 3000

        btn_login.startAnimation(btnLoginAnimation)

        val btnRegisterStartPosition = layout_button_container.width - btn_register.x
        val btnRegisterEndPosition = 0f

        val btnRegisterTranslate = TranslateAnimation(Animation.ABSOLUTE,
                btnRegisterStartPosition,
                Animation.ABSOLUTE,
                btnRegisterEndPosition,
                Animation.RELATIVE_TO_SELF,
                0f,
                Animation.RELATIVE_TO_SELF,
                0f)

        val btnRegisterRotate = RotateAnimation(0f,
                -360f,
                Animation.RELATIVE_TO_SELF,
                0.5f,
                Animation.RELATIVE_TO_SELF,
                0.5f)

        val btnRegisterAnimation = AnimationSet(true)
        btnRegisterAnimation.addAnimation(btnRegisterRotate)
        btnRegisterAnimation.addAnimation(btnRegisterTranslate)
        btnRegisterAnimation.duration = 3000

        btn_register.startAnimation(btnRegisterAnimation)
    }

    /**
     * 通过XML Resource实现View动画，按钮起始位置和结束位置是不精确的
     */
    private fun doButtonsViewAnimationFromRes() {
        val animRollInFromLeft = AnimationUtils.loadAnimation(applicationContext, R.anim.anim_roll_in_from_left)
        val animRollInFromRight = AnimationUtils.loadAnimation(applicationContext, R.anim.anim_roll_in_from_right)
        btn_register.startAnimation(animRollInFromRight)
        btn_login.startAnimation(animRollInFromLeft)
    }
}
